src/middlewares/balance.ts

```diff
@@ -174,23 +178,25 @@ export default class BalanceMiddleware implements Middleware {
               return next(data)
             }
 
-            // We do nothing here (i.e. unlike for incoming packets) and wait until the packet is fulfilled
-            // This means we always take the most conservative view of our balance with the upstream peer
             let result
             try {
+
+              balance.subtract(parsedPacket.amount)
```

Comment:
If we subtract on an outgoing packet and then add on a reject, it could put us above our maximum balance.

Imagine my max balance is 100, current balance is 80.

1. I send out a packet for 30. Now my balance is 50 because outgoing packet was subtracted on prepare.
2. I get an incoming packet for 40. Now my balance is 90 because incoming packets are added.
3. Outgoing packet for 30 is rejected, meaning that it's added back to the balance. Now our balance is 120 which is greater than our maximum.

The only correct way I can think of is to track two balances, one that represents your minimum balance possible and one that represents your maximum balance possible. Outgoing packets are counted against min-possible but incoming packets are not added. Incoming packets are counted against max-possible but outgoing packets are not subtracted. min-possible can never exceed min balance and max-possible can never exceed max balance.